### PR TITLE
KEYCLOAK-19571 Add indices to HotRodClientEntity fields

### LINKED ISSUE #9240: Add indexing to HotRodGroupEntity
### Description

We need to add indexing to groups similarly as for clients: https://github.com/keycloak/keycloak/blob/2266ab4c15e71a504d43a242dcf951d3673e3099/model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/client/HotRodClientEntity.java#L41

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by martin-kanis**
This PR adds indices on `HotRodClientEntity` fields. It also bumps Infinispan version to 12 because of Infinispan 12 is able to define an index in cache declaration without the need to put proto schema into metadata cache first. Therefore we should first merge https://github.com/keycloak/keycloak/pull/9037 which also bumps the Infinispan version. 

Closes https://issues.redhat.com/browse/KEYCLOAK-19571
Closes https://github.com/keycloak/keycloak/issues/9240

### CODE REVIEW

File: core/src/main/java/org/keycloak/Config.java [hunk 1]
```diff
@@ -91,7 +91,7 @@ public String get(String key) {
         @Override
         public String get(String key, String defaultValue) {
             String v = System.getProperty(prefix + key, defaultValue);
-            return v != null && !v.isEmpty() ? v : null;
+            return v != null && !v.isEmpty() ? v : defaultValue;
```
Comment (hmlnarik) [inline core/src/main/java/org/keycloak/Config.java:94]: Please revert this change. Either this is a bug that should be treated separately, or a workaround that should be treated properly.

Comment (martin-kanis) [reply core/src/main/java/org/keycloak/Config.java]: Reverted

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/ClientModelTest.java [hunk 2]
```diff
@@ -76,6 +90,47 @@ public void testClientsBasics() {
...
+                return null;
+            });
+            assertThat(id.get(), is(equalTo(searchClient.getId())));
+            assertThat(clientId.get(), is(equalTo(searchClientId)));
+
+            withRealm(realmId, (session, realm) -> {
+                getClientIds(session.clients().searchClientsByClientIdStream(realm, "sp*ci_l Ch***cters", 0, 10), id, clientId);
+                return null;
+            });
+            assertThat(id.get(), is(equalTo(searchClient.getId())));
```
Comment (mhajas) [inline testsuite/model/src/test/java/org/keycloak/testsuite/model/ClientModelTest.java:109]: Just a nitpick: it would be easier to move `assertThat` into the lambda in the `withRealm` call. This way you would avoid setting `id` and `clientId` to the `AtomicRefence`.

**Review by mhajas — COMMENTED** [review]
Thank you for the PR @martin-kanis! I added a few comments into the code.

**Review by mhajas — COMMENTED** [review]
Thanks for the changes @martin-kanis, I added a few more comments

File: pom.xml [hunk 3]
```diff
@@ -78,8 +78,8 @@
         <jakarta.persistence.version>2.2.3</jakarta.persistence.version>
         <hibernate.core.version>5.3.20.Final</hibernate.core.version>
         <hibernate.c3p0.version>5.3.20.Final</hibernate.c3p0.version>
-        <infinispan.version>11.0.9.Final</infinispan.version>
-        <infinispan.protostream.processor.version>4.3.4.Final</infinispan.protostream.processor.version>
+        <infinispan.version>12.1.7.Final</infinispan.version>
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply pom.xml:81]: Here it is: https://github.com/keycloak/keycloak/pull/9333 See description for explanation.

**Comment by martin-kanis** [comment]
This PR requires https://github.com/keycloak/keycloak/pull/9333 to be merged first.

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryMapModelCriteriaBuilder.java [hunk 4]
```diff
@@ -175,6 +200,67 @@ private StringBuilder getWhereClauseBuilder() {
...
 
+    public static Object getFirstArrayElement(Object[] value) throws IllegalStateException {
+        if (value == null || value.length == 0) {
+            throw new IllegalStateException("Invalid argument: " + Arrays.toString(value));
+        }
+        return value[0];
+    }
+
+    public static Object[] sanitizeValues(SearchableModelField<?> modelField, Operator op, Object[] values) {
+        if (isAnalyzedModelField(modelField) || isOverriddenModelField(modelField, op)) {
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryMapModelCriteriaBuilder.java]: No it doesn't work for example with operators `EQ` or `IN` if a string is sanitized. A group with `name = "my_group_name"` searched by sanitized string `my\\_group\\_name` is not matched as it looks for exact match. In other words, those special characters are not expanded unless `LIKE` operator is used.

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryMapModelCriteriaBuilder.java [hunk 5]
```diff
@@ -175,6 +189,43 @@ private StringBuilder getWhereClauseBuilder() {
         return whereClauseBuilder;
     }
 
+    public static Object sanitize(Object value) {
```
Comment (mhajas) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryMapModelCriteriaBuilder.java:192]: This method is unused.

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryOperators.java [hunk 6]
```diff
@@ -74,25 +76,31 @@
...
          * Produces an Ickle query where clause for obtained parameters
          *
+         * @param modelField searchable field
          * @param fieldName left side operand
          * @param values right side operands
          * @param parameters mapping between named parameters and actual parameter values
          * @return resulting string that will be part of resulting
          */
-        String combine(String fieldName, Object[] values, Map<String, Object> parameters);
+        String combine(SearchableModelField<?> modelField, String fieldName, Object[] values, Map<String, Object> parameters);
```
Comment (mhajas) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryOperators.java:85]: We introduced this parameter to distinguish between analyzed and non-analyzed fields. However, currently, we are not using it, should we remove this parameter?

**Review by mhajas — CHANGES_REQUESTED** [review]
Nice work @martin-kanis! I added just a few final comments. Also, have you seen the GHA failures? It seems that there is something going on with Infinispan tests, is it caused by these changes?

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryMapModelCriteriaBuilder.java (hunk 5 above)
Comment (martin-kanis) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryMapModelCriteriaBuilder.java:192]: Good catch. It should be called from `iLike` method.

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryOperators.java (hunk 6 above)
Comment (martin-kanis) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryOperators.java]: yes, removed.

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryOperators.java [hunk 7]
```diff
@@ -82,17 +84,29 @@
...
-        String field = C + "." + modelField;
+    private static String notExists(String modelFieldName, Object[] values, Map<String, Object> parameters) {
+        String field = C + "." + modelFieldName;
         return field + " IS NULL OR " + field + " IS EMPTY";
     }
 
-    private static String in(String modelField, Object[] values, Map<String, Object> parameters) {
+    private static String iLike(String modelFieldName, Object[] values, Map<String, Object> parameters) {
+        String sanitizedValue = (String) IckleQueryMapModelCriteriaBuilder.sanitize(values[0]);
+        return singleValueOperator(ModelCriteriaBuilder.Operator.ILIKE)
```
Comment (mhajas) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryOperators.java:99]: This smells with a possibility to introduce an infinite loop in the future.
```suggestion
        return singleValueOperator(ModelCriteriaBuilder.Operator.LIKE)
```

**Review by mhajas — APPROVED** [review]
Thank you for fixes @martin-kanis. I found one more nice-to-have thing, but I am leaving this to you, maybe we can replace it later. Otherwise, it looks good to me. Sorry for the late reviews.

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryOperators.java (hunk 7 above)
Comment (martin-kanis) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryOperators.java:99]: Why do you think so? In `singleValueOperator` the ILIKE operator is exchanged for "LIKE" string using `IckleQueryOperators.operatorToString(op)`.

Comment (mhajas) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryOperators.java:99]: You are correct. Currently, this works like this. What this method does is that it delegates the creation of the resulting query to another operator (LIKE) with a different parameter - lowercase. However, it is now using `ILIKE`, which works because `ILIKE` is mapped to the same operator string. 

I just find it a little bit confusing that `ILIKE` is used again. This is just a nitpick so feel free to ignore it.

**Review by hmlnarik — APPROVED** [review]
Approving per @mhajas 's approval.
